File: Sean的leetcode/Leetcode/Maximal Square.cpp
//https://leetcode.com/discuss/38489/easy-solution-with-detailed-explanations-8ms-time-and-space
//O(max_row*max_col) time, O(max_col) space.
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int max_row = matrix.size();

        if (max_row == 0) return 0;

        int max_col = matrix[0].size();

        //the maximal size of the square that can be achieved at col(i) which is the previous row.
        vector<int>pre_row(max_col, 0);
        int max_size = 0;
        int curr_size;
        int left_size;
        for (int i = 0; i < max_row; i++)
            for (int j = 0; j < max_col; j++)
            {
                //first row/col: matrix[i][j] is 0: square size is 0.
                //               matrix[i][j] is 1: square size is 1.
                if (i == 0 || j == 0)
                    curr_size = matrix[i][j] - '0';
                else
                {
                    if (matrix[i][j] == '1')
                    {
                        curr_size = min(min(left_size, pre_row[j]), pre_row[j-1]) + 1;
                    }
                    else
                        curr_size = 0;

                    //update pre_row[j-1] to the previous curr_size
                    pre_row[j-1] = left_size;

                }

                //if j is the last col or i is the first row, update curr_size directly
                if (j == max_col - 1 || i == 0)
                    pre_row[j] = curr_size;

                max_size = max(max_size, curr_size);
                left_size = curr_size;
            }
        return max_size * max_size;
    }
};

File: Sean的leetcode/python/Maximal Square.py
###
# DP
# Time Complexity: O(m*n)
# Space Complexity: O(n)
# P[i][j]:the maximal size of the square that can be achieved at point (i, j)
# i == 0 or j == 0: P[i][j] = matrix[i][j]
# i > 0 and j > 0: if matrix[i][j] = 0 -> P[i][j] = 0
#                 if matrix[i][j] = 1 -> P[i][j] = min(P[i-1][j], P[i][j-1], P[i-1][j-1]) + 1 (three neighbors must be 1)
###    
### Solution 1
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if not matrix or not matrix[0]:
            return 0
        dp = [0] * len(matrix[0])
        max_side = 0
        for i in xrange(len(matrix)):
            new_dp = [0] * len(matrix[0])
            for j in xrange(len(matrix[0])):
                if j == 0:
                    new_dp[j] = int(matrix[i][j])
                else:
                    if matrix[i][j] == '1':
                        new_dp[j] = min(dp[j-1], dp[j], new_dp[j-1]) + 1
                max_side = max(max_side, new_dp[j])
            dp = new_dp
        return max_side*max_side
                        
### Solution 2
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if not matrix or not matrix[0]:
            return 0
        
        max_size = 0
        left = 0
        dp = [0] * len(matrix[0])
        for i in xrange(len(matrix)):
            for j in xrange(len(matrix[0])):
                if i == 0 or j == 0:
                    curr = int(matrix[i][j])
                else:
                    if matrix[i][j] == '1':
                        curr = min(left, dp[j-1], dp[j]) + 1
                    else:
                        curr = 0
                max_size = max(max_size, curr)
                if j > 0:
                    dp[j-1] = left
                if j == len(matrix[0])-1:
                    dp[j] = curr
                left = curr
        return max_size * max_size